### PR TITLE
Complete seq2seq for fluid

### LINKED ISSUE #22: Need add seq2seq model based on fluid
- [x] https://github.com/PaddlePaddle/Paddle/issues/6764
- [ ] https://github.com/PaddlePaddle/Paddle/issues/6757
- [x] https://github.com/PaddlePaddle/Paddle/issues/6712
- [x] https://github.com/PaddlePaddle/Paddle/issues/6590
- [x] https://github.com/PaddlePaddle/Paddle/issues/6581
- [x] https://github.com/PaddlePaddle/Paddle/issues/7173
- [x] https://github.com/PaddlePaddle/Paddle/issues/7195  Add static_input for DynamicRNN.
- [x] https://github.com/PaddlePaddle/Paddle/pull/7540 Reorder the decoder_boot.
- [x] https://github.com/PaddlePaddle/Paddle/pull/7363 Fix sequence_pool.

### LINKED ISSUE #55: Need complete seq2seq for fluid

### PR DESCRIPTION
**Pull request by pkuyym**
Resolves #55 
Resolves #22

### CODE REVIEW

File: fluid/machine_translation.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+parser.add_argument(
+    "--use_gpu",
+    type=distutils.util.strtobool,
+    default=True,
+    help="Whether use gpu. (default: %(default)d)")
+parser.add_argument(
+    "--max_length",
+    type=int,
+    default=250,
+    help="The max length of sequence when doing generation. "
```
Comment (ranqiu92) [inline fluid/machine_translation.py]: max -> maximum

File: fluid/machine_translation.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+parser.add_argument(
+    "--decoder_size",
+    type=int,
+    default=512,
+    help="The size of decoder rnn unit. (default: %(default)d)")
+parser.add_argument(
+    "--batch_size",
+    type=int,
+    default=16,
+    help="The sequence number of a batch data. (default: %(default)d)")
```
Comment (ranqiu92) [inline fluid/machine_translation.py]: of a mini-batch

File: fluid/machine_translation.py [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+parser.add_argument(
+    "--word_vector_dim",
+    type=int,
+    default=512,
+    help="The dimension of embedding table. (default: %(default)d)")
+parser.add_argument(
+    "--encoder_size",
+    type=int,
+    default=512,
+    help="The size of encoder bi-rnn unit. (default: %(default)d)")
```
Comment (ranqiu92) [inline fluid/machine_translation.py]: size -> dimension

File: fluid/machine_translation.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+parser.add_argument(
+    "--decoder_size",
+    type=int,
+    default=512,
+    help="The size of decoder rnn unit. (default: %(default)d)")
```
Comment (ranqiu92) [inline fluid/machine_translation.py]: size -> dimension

File: fluid/machine_translation.py [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+parser.add_argument(
+    "--beam_size",
+    type=int,
+    default=3,
+    help="The width for beam searching. (default: %(default)d)")
+parser.add_argument(
+    "--use_gpu",
+    type=distutils.util.strtobool,
+    default=True,
+    help="Whether use gpu. (default: %(default)d)")
```
Comment (ranqiu92) [inline fluid/machine_translation.py]: to use

File: fluid/machine_translation.py [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+        input=src_reversed, pool_type='first')
+
+    decoder_boot = fluid.layers.fc(input=backward_first,
+                                   size=decoder_size,
+                                   bias_attr=False,
+                                   act='tanh')
+
+    def lstm_decoder_with_attention(target_embedding, encoder_vec, encoder_proj,
+                                    decoder_boot, decoder_size):
+        def simple_attention(encoder_vec, encoder_proj, decoder_state):
```
Comment (ranqiu92) [inline fluid/machine_translation.py]: The attention mechanism is wrong. Where is 'tanh' operation which appears in original formula?

Comment (pkuyym) [reply fluid/machine_translation.py]: Didn't catch your point, why `tanh` is necessary for attention? There are several kind of attention mechanisms. Please refer to https://github.com/PaddlePaddle/Paddle/blob/9bfa3013891cf3da832307894acff919d6705cee/python/paddle/trainer_config_helpers/networks.py#L1400

Comment (ranqiu92) [reply fluid/machine_translation.py]: https://github.com/PaddlePaddle/Paddle/blob/9bfa3013891cf3da832307894acff919d6705cee/python/paddle/trainer_config_helpers/networks.py#L1473   
Here, the mixed_layer performs  `tanh`.  
If you realizes the attention mechanism in` Neural Machine Translation By Jointly Learning To Align and Translate`, tanh is needed.

Comment (pkuyym) [reply fluid/machine_translation.py]: Why do you think it's wrong to apply linear transformation?

Comment (pkuyym) [reply fluid/machine_translation.py]: To keep consistent, will apply `tanh` in next PR. Thanks.

Comment (ranqiu92) [reply fluid/machine_translation.py]: Maybe we shouldn't focus on one word. And is the original attention in that paper what you want to realize? Or a modified one?

File: fluid/machine_translation.py [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+                   max_length=250):
+    """Construct a seq2seq network."""
+    feeding_list = ["source_sequence", "target_sequence", "label_sequence"]
+
+    def bi_lstm_encoder(input_seq, size):
+        input_forward_proj = fluid.layers.fc(input=input_seq,
+                                             size=size * 4,
+                                             act='tanh')
+        forward, _ = fluid.layers.dynamic_lstm(
+            input=input_forward_proj, size=size * 4)
```
[the remark this answers is not in the document]
Comment (pkuyym) [reply fluid/machine_translation.py]: > double check dynamic_lstm need 4. I roughly remember it has been done inside the lstm layer.

https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/fluid/layers/nn.py#L231

> name a gate_size is a good idea.

Agree.

File: fluid/machine_translation.py [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+    type=int,
+    default=512,
+    help="The size of decoder rnn unit. (default: %(default)d)")
+parser.add_argument(
+    "--batch_size",
+    type=int,
+    default=16,
+    help="The sequence number of a batch data. (default: %(default)d)")
+parser.add_argument(
+    "--dict_size",
```
[the remark this answers is not in the document]
Comment (pkuyym) [reply fluid/machine_translation.py]: Seems this is an argument:

https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/dataset/wmt14.py#L106

File: fluid/machine_translation.py [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+import paddle.v2 as paddle
+import paddle.v2.fluid as fluid
+import paddle.v2.fluid.core as core
+import paddle.v2.fluid.framework as framework
+from paddle.v2.fluid.param_attr import ParamAttr
+from paddle.v2.fluid.executor import Executor
+
+parser = argparse.ArgumentParser(description=__doc__)
+parser.add_argument(
+    "--word_vector_dim",
```
[the remark this answers is not in the document]
Comment (pkuyym) [reply fluid/machine_translation.py]: Followed.

File: fluid/machine_translation.py (hunk 3 above)
Comment (pkuyym) [reply fluid/machine_translation.py]: Thinks both are ok, but size is shorter.

File: fluid/machine_translation.py (hunk 4 above)
Comment (pkuyym) [reply fluid/machine_translation.py]: Thinks both are ok, but size is shorter.

File: fluid/machine_translation.py (hunk 2 above)
Comment (pkuyym) [reply fluid/machine_translation.py]: Followed.

File: fluid/machine_translation.py [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+    default=16,
+    help="The sequence number of a batch data. (default: %(default)d)")
+parser.add_argument(
+    "--dict_size",
+    type=int,
+    default=30000,
+    help="The dictionary capacity. Dictionaries of source sequence and "
+    "target dictionary have same capacity. (default: %(default)d)")
+parser.add_argument(
+    "--pass_number",
```
[the remark this answers is not in the document]
Comment (pkuyym) [reply fluid/machine_translation.py]: Followed.

File: fluid/machine_translation.py [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+    "--learning_rate",
+    type=float,
+    default=0.0002,
+    help="Learning rate used to train the model. (default: %(default)f)")
+parser.add_argument(
+    "--mode",
+    type=str,
+    default='train',
+    choices=['train', 'infer'],
+    help="Do training or inference. (default: %(default)s)")
```
[the remark this answers is not in the document]
Comment (pkuyym) [reply fluid/machine_translation.py]: Thanks, followed.

File: fluid/machine_translation.py (hunk 5 above)
Comment (pkuyym) [reply fluid/machine_translation.py]: Thanks, followed.

File: fluid/machine_translation.py (hunk 1 above)
Comment (pkuyym) [reply fluid/machine_translation.py]: Followed.

File: fluid/machine_translation.py [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+
+
+def seq_to_seq_net(word_vector_dim,
+                   encoder_size,
+                   decoder_size,
+                   source_dict_dim,
+                   target_dict_dim,
+                   is_generating=False,
+                   beam_size=3,
+                   max_length=250):
```
[the remark this answers is not in the document]
Comment (pkuyym) [reply fluid/machine_translation.py]: Followed.

File: fluid/machine_translation.py [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+    """Construct a seq2seq network."""
+    feeding_list = ["source_sequence", "target_sequence", "label_sequence"]
+
+    def bi_lstm_encoder(input_seq, size):
```
[the remark this answers is not in the document]
Comment (pkuyym) [reply fluid/machine_translation.py]: Add detailed comments.

File: fluid/machine_translation.py [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+        total_loss = 0.0
+        count = 0
+        for batch_id, data in enumerate(test_batch_generator()):
+            src_seq = to_lodtensor(map(lambda x: x[0], data), place)[0]
+            trg_seq = to_lodtensor(map(lambda x: x[1], data), place)[0]
+            lbl_seq = to_lodtensor(map(lambda x: x[2], data), place)[0]
+
+            fetch_outs = exe.run(
+                inference_program,
+                feed=dict(zip(*[feeding_list, (src_seq, trg_seq, lbl_seq)])),
```
Comment (dzhwinter) [inline fluid/machine_translation.py]: Please fix this issue. 
even dict is better than `*zip()` function sugar.

**Pull request merged**